massa-models/src/denunciation.rs

```diff
@@ -258,6 +246,35 @@ impl Denunciation {
             && public_key.verify_signature(&hash_1, &signature_1).is_ok()
             && public_key.verify_signature(&hash_2, &signature_2).is_ok())
     }
+
+    /// Get Denunciation slot ref
+    pub fn get_slot(&self) -> &Slot {
+        match self {
+            Denunciation::Endorsement(endo_de) => &endo_de.slot,
+            Denunciation::BlockHeader(blkh_de) => &blkh_de.slot,
+        }
+    }
+
+    /// For a given slot (and given the slot at now()), check if it can be denounced
+    /// Can be used to check if block header | endorsement is not too old (at reception or too cleanup cache)
+    pub fn is_expired(
+        slot: &Slot,
+        last_cs_final_periods: &[u64],
+        periods_per_cycle: u64,
+        denunciation_expire_cycle_delta: u64,
+    ) -> bool {
+        // If the Slot is final, a Denunciation can still be made for 1 cycle
+        if slot.period <= last_cs_final_periods[slot.thread as usize] {
+            let cycle_of_slot = slot.get_cycle(periods_per_cycle);
+
+            let slot_final = Slot::new(last_cs_final_periods[slot.thread as usize], slot.thread);
+            let cycle_of_final = slot_final.get_cycle(periods_per_cycle);
+
+            (cycle_of_final - cycle_of_slot) > denunciation_expire_cycle_delta
```

Comment:
let's count in periods instead of cycles so this whole function becomes:

```
is_expired(...) -> bool {
   last_cs_final_periods[slot.thread].checked_sub(slot.period) > Some(denunciation_expire_periods)
}

```